### PR TITLE
Convert RAJA, CHAI and Umpire to CachedCMakePackages

### PR DESCRIPTION
**Pull request by davidbeckingsale**
Closes #25484

### CODE REVIEW

File: var/spack/repos/builtin/packages/chai/package.py [hunk 1]
```diff
@@ -85,60 +87,79 @@ class Chai(CMakePackage, CudaPackage, ROCmPackage):
 
     conflicts('+benchmarks', when='~tests')
 
-    def cmake_args(self):
+    def _get_sys_type(self, spec):
+        sys_type = spec.architecture
```
Comment (adrienbernede) [inline var/spack/repos/builtin/packages/chai/package.py:91]: What is the difference with using self.spec?

File: var/spack/repos/builtin/packages/chai/package.py [hunk 2]
```diff
@@ -85,60 +87,79 @@ class Chai(CMakePackage, CudaPackage, ROCmPackage):
...
+            entries.append(cmake_cache_option("CMAKE_CUDA_SEPARABLE_COMPILATION", True))
+            entries.append(cmake_cache_option("CUDA_SEPARABLE_COMPILATION", True))
 
             if not spec.satisfies('cuda_arch=none'):
                 cuda_arch = spec.variants['cuda_arch'].value
-                options.append('-DCUDA_ARCH=sm_{0}'.format(cuda_arch[0]))
+                entries.append(cmake_cache_string(
+                    "CUDA_ARCH", 'sm_{0}'.format(cuda_arch[0])))
+                entries.append(cmake_cache_string(
+                    "CMAKE_CUDA_ARCHITECTURES={0}".format(cuda_arch[0])))
```
Comment (adrienbernede) [inline var/spack/repos/builtin/packages/chai/package.py]: Is that the right syntax? Above and below we see `cmake_cache_string` taking 2 arguments.

**Comment by davidbeckingsale** [comment]
Can I get some help with the GitLab again? I see this:

```
spack.error.SpackError: No installed spec matches the hash: 'w33kbsvghm6a3eiyq35n5lpeda6hgjqk'
```

but I'm not sure how these changes would cause that error.

**Review by adrienbernede — COMMENTED** [review]
Apparently I never clicked "submit"...

**Comment by adrienbernede** [comment]
Hey @davidbeckingsale, these errors are generally not the real ones, but rather a final message.
the real error message is:
```
==> [2021-09-03-20:59:08.017007] Error: TypeError: cmake_cache_string() missing 1 required positional argument: 'value'
/builds/spack/spack/var/spack/repos/builtin/packages/raja/package.py:105, in initconfig_hardware_entries:
        102                entries.append(cmake_cache_string(
        103                    "CUDA_ARCH", 'sm_{0}'.format(cuda_arch[0])))
        104                entries.append(cmake_cache_string(
  >>    105                    "CMAKE_CUDA_ARCHITECTURES={0}".format(cuda_arch[0])))
        106        else:
        107            entries.append(cmake_cache_option("ENABLE_CUDA", False))
        108
```

And it seems this is due to a syntax error I mentioned in my review, which I forgot to "submit"...

**Comment by davidbeckingsale** [comment]
Ah perfect, thank you Adrien - those are my fault and I can fix them :D

**Comment by adrienbernede** [comment]
A very interesting type of failure now... especially since I don't see how your last commit could have changed anything.
Maybe a change in spack@develop in the meantime?

I think we won the right to ping @scottwittenburg.

**Comment by scottwittenburg** [comment]
Yes, the gitlab pipeline failure doesn't have anything to do with this PR, other than this PR generated a job for a full hash that failed recently on develop.  We do this in the spirit of trying to avoid a lot of wasteful attempts to build something known to fail.  In this case, my suspicion is that the `kokkos-kernels` failure on develop was due to running on a cloud runner without enough resources to successfully complete the build.

Once I resolve that issue (it might resolve itself too, but maybe I can speed it along), we can ask spackbot to re-run this pipeline.

**Comment by adrienbernede** [comment]
> Yes, the gitlab pipeline failure doesn't have anything to do with this PR, other than this PR generated a job for a full hash that failed recently on develop. We do this in the spirit of trying to avoid a lot of wasteful attempts to build something known to fail. In this case, my suspicion is that the `kokkos-kernels` failure on develop was due to running on a cloud runner without enough resources to successfully complete the build.
> 
> Once I resolve that issue (it might resolve itself too, but maybe I can speed it along), we can ask spackbot to re-run this pipeline.

Sounds like a plan, thanks!

**Comment by scottwittenburg** [comment]
@spackbot run pipeline

**Comment by adrienbernede** [comment]
@scottwittenburg I wanted to seize the occasion to say that apart from the issues we reported in https://github.com/spack/spack/pull/25528 (related to the concretizer), I find this pipeline setup pleasantly responsive and effective, with seamless navigation.

**Review by adrienbernede — APPROVED** [review]

**Pull request merged**